File: LeetCode/Medium/1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        dict_s = {}
        for char in s:
            if char in dict_s:
                dict_s[char] += 1
            else:
                dict_s[char] = 1
        
        sorted_s = sorted(dict_s.items(), key=lambda x: x[1])

        cnt_set = set()
        prev_cnt = 0
        ans = 0
        for char, cnt in sorted_s:
            if cnt in cnt_set:
                sub_cnt = 0
                while cnt in cnt_set:
                    cnt -= 1
                    sub_cnt += 1
                    if cnt <=0:
                        break
                ans += sub_cnt
                cnt_set.add(cnt)
            else:
                cnt_set.add(cnt)
        return ans 

File: LeetCode/Medium/1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        map<char, int> dict_s;
        for(char c : s) {
            dict_s[c]++;
        }

        vector<pair<char, int>> sorted_s(dict_s.begin(), dict_s.end());
        sort(sorted_s.begin(), sorted_s.end(), [](const pair<char, int>& a, const pair<char, int>& b){
            return a.second < b.second;
        });

        set<int> cnt_set;
        int ans = 0;
        for(auto &[char_, cnt] : sorted_s) {
            if(cnt_set.find(cnt) != cnt_set.end()) {
                int sub_cnt = 0;
                while(cnt_set.find(cnt) != cnt_set.end()) {
                    cnt--;
                    sub_cnt++;
                    if(cnt <= 0) {
                        break;
                    }
                }
                ans += sub_cnt;
                if(cnt > 0) {
                    cnt_set.insert(cnt);
                }
            } else {
                cnt_set.insert(cnt);
            }
        }
        return ans;
    }
};